accepted/future-releases/static-extension-methods/design-document.md

```diff
@@ -0,0 +1,501 @@
+
+If the inference fails, or if the synthetic constructor invocation (`MyList(times)` in the above example) would not be statically valid for any other reason, then the extension does not apply. If an extension does not have an instance member with the base name `quickSort`, it does not apply.
+
+If exactly one extension applies to an otherwise failed member invocation, then that invocation becomes an invocation of the corresponding instance member of the extension, with `this` bound to the receiver object and extension type parameters bound to the inferred types.
+
+If the member is itself generic and has no type parameters supplied, normal static type inference applies again.
+
+It is *as if* the invocation `times.quickSort` was converted to `MyList<Duration>(times).quickSort()`. The difference is that there is never an actual `MyList` object, and the `this` object inside `quickSort` is just the `times` list itself (although the extension methods apply to that code too).
+
+### Extension Conflict Resolution
+
+If more than one extension applies to a specific member invocation, then if exactly one of them is more specific than all the others, that one is chosen. Otherwise it is a compile-time error.
+
+An extension with `on` type clause *T*<sub>1</sub> is more specific than another extension with `on` type clause *T*<sub>2</sub> iff the instantiated type (the type after applying type inference from the receiver) of *T*<sub>1</sub> is a subtype of the instantiated type of *T*<sub>2</sub> and either:
+
+- not vice versa, or
+- the instantiate-to-bounds type of *T*<sub>1</sub> is a subtype of the instantiate-to-bounds type of *T*<sub>2</sub> and not vice versa.
+
+This definition is designed to ensure that the extension chosen is the one that has the most precise type information available.
+
+If an extension's `on` type has a trailing `?`, say `T?`, then we treat it just as we would for the corresponding nullable type with NNBD types, which are described in a separate design document. In short, it means treating the type `T?`, for subtyping purposes, as a union type of `T` with `Null` (a least supertype of `T` and `Null`).
+
+That is, the specificity of an extension wrt. an application depends of the type it is used at, and how specific the extension is itself (what its implementation can assume about the type). 
+
+Example:
+
+```dart
+extension SmartIterable<T> on Iterable<T> {
+  void doTheSmartThing(void Function(T) smart) {
+    for (var e in this) smart(e);
+  }
+}
+extension SmartList<T> on List<T> {
+  void doTheSmartThing(void Function(T) smart) {
+    for (int i = 0; i < length; i++) smart(this[i]);
+  }
+}
+...
+  List<int> x = ....;
+  x.doTheSmartThing(print);
+```
+
+Here both the extensions apply, but the `SmartList` extension is more specific than the `SmartIterable` extension because `List<dynamic>` &lt;: `Iterable<dynamic>`.
+
+Example:
+
+```dart
+extension BoxCom<T extends num> on Box<Iterable<T>> { T best() {...} }
```

Comment:
did you mean `on Iterable<T>` (remove the surrounding `Box<..>` here and below? or change teh code further down to say `Box<List<int>> x = ...`?)

Reply:
Not sure what I intended, but the `Box` type is unnecessary for the example, so I'll remove it.